delfin/drivers/pure/flasharray/pure_flasharray.py

```diff
@@ -363,6 +364,134 @@ def reset_connection(self, context, **kwargs):
+        return metrics
+
+    def get_metrics(self, storage_id, resource_metrics, start_time, end_time,
+                    url, resource_type, resource_id=None, resource_name=None):
+        metrics = []
+        localtime = time.mktime(time.localtime())
+        localtime_ms = localtime * units.k
+        if start_time < localtime_ms - consts.ONE_YEAR_DIFFERENCE\
+                or end_time < start_time or start_time > localtime_ms:
+            return metrics
+        historical = self.get_historical(localtime_ms, start_time)
+        list_metrics = self.rest_handler.rest_call(url.format(historical))
+        for storage_metrics in (list_metrics or []):
+            opened = storage_metrics.get('time')
+            time_difference = localtime - time.mktime(time.gmtime())
+            timestamp = (int(datetime.datetime.strptime
+                             (opened, '%Y-%m-%dT%H:%M:%SZ').timestamp()
+                             + time_difference) *
+                         consts.DEFAULT_LIST_ALERTS_TIME_CONVERSION) if\
+                opened is not None else None
+            if timestamp < start_time or timestamp > end_time:
+                continue
+            read_iop = storage_metrics.get('reads_per_sec')
+            write_iop = storage_metrics.get('writes_per_sec')
+            read_throughput = storage_metrics.get('output_per_sec') / units.Mi
+            write_throughput = storage_metrics.get('input_per_sec') / units.Mi
+            response_time = \
+                (storage_metrics.get('usec_per_read_op')
+                 + storage_metrics.get('usec_per_write_op'))\
+                / units.k / consts.AVERAGE_TWO
+            metrics_data = {
+                'iops': read_iop + write_iop,
+                "readIops": read_iop,
+                "writeIops": write_iop,
+                "throughput": read_throughput + write_throughput,
+                "readThroughput": read_throughput,
+                "writeThroughput": write_throughput,
+                "responseTime": response_time,
+            }
+            for resource_key in resource_metrics.keys():
+                if resource_type == constants.ResourceType.VOLUME:
+                    resource_name = resource_id = storage_metrics.get('name')
+                labels = {
+                    'storage_id': storage_id,
+                    'resource_type': resource_type,
+                    'resource_id': resource_id,
+                    'resource_name': resource_name,
+                    'type': 'RAW',
+                    'unit': resource_metrics[resource_key]['unit']
+                }
+                resource_value = {timestamp: metrics_data.get(resource_key)}
+                metrics_res = constants.metric_struct(name=resource_key,
+                                                      labels=labels,
+                                                      values=resource_value)
+                metrics.append(metrics_res)
```

Comment:
This method code is too long

Reply:
Has been modified

metrics_res = constants.metric_struct(
                    name=resource_key, labels=labels, values=resource_value)